### PR TITLE
chore: Fix save behavior for Django 3 and drop Django 2 support

### PR DESCRIPTION
**Pull request by bgrant**
Work related to this change in Django 3:

[Django 3.0 release notes | Django documentation | Django](https://docs.djangoproject.com/en/4.0/releases/3.0/#model-save-when-providing-a-default-for-the-primary-key)

### CODE REVIEW

File: django_declarative_apis/machinery/__init__.py [hunk 1]
```diff
@@ -907,6 +907,11 @@ class ResourceEndpointDefinition(EndpointDefinition):
     :code:`ResourceEndpointDefinition` is mainly used for :code:`GET`.
     """
 
+    def __init__(self, *args, **kwargs):
```
Comment (bgrant) [inline django_declarative_apis/machinery/__init__.py:910]: There's already an `__init__` further down - I assume we should merge these @dshafer ?

File: django_declarative_apis/machinery/__init__.py [hunk 2]
```diff
@@ -213,13 +244,18 @@ def get_response(self):  # noqa: C901
                     immediate_task.run(self.bound_endpoint)
 
             except errors.ClientError as ce:
-                if ce.save_changes and resource and resource.is_dirty():
-                    resource.save()
+                if (
+                    ce.save_changes
+                    and resource
+                    and hasattr(resource, "is_dirty")
```
Comment (bgrant) [inline django_declarative_apis/machinery/__init__.py:250]: This was a missing `hasattr` which allows DDA to work with models that don't use DirtyFields

File: django_declarative_apis/machinery/__init__.py [hunk 3]
```diff
@@ -904,10 +940,10 @@ class ResourceEndpointDefinition(EndpointDefinition):
     """
 
     def __init__(self, *args, **kwargs):
-        super().__init__()
+        super().__init__(*args, **kwargs)
```
Comment (bgrant) [inline django_declarative_apis/machinery/__init__.py:943]: This is a fix noticed in passing (unrelated to Django 3)

File: django_declarative_apis/machinery/__init__.py (hunk 2 above)
Comment (bgrant) [reply django_declarative_apis/machinery/__init__.py:250]: (unrelated to Django 3)

File: django_declarative_apis/machinery/__init__.py [hunk 4]
```diff
@@ -101,9 +103,11 @@ def get_instance_value(self, owner_instance, owner_class):
         try:
             value = self.func(owner_instance)
         except django.core.exceptions.ObjectDoesNotExist:
-            raise errors.ClientErrorNotFound(
-                "{0} instance not found".format(self.type.__name__)
-            )
+            if hasattr(self.type, "__name__"):
+                message = "{0} instance not found".format(self.type.__name__)
```
Comment (demianbrecht) [inline django_declarative_apis/machinery/__init__.py:107]: f-string would be preferred here, no? or are we concerned about preserving python backwards compatibility?

File: django_declarative_apis/machinery/__init__.py [hunk 5]
```diff
@@ -101,9 +103,11 @@ def get_instance_value(self, owner_instance, owner_class):
         try:
             value = self.func(owner_instance)
         except django.core.exceptions.ObjectDoesNotExist:
-            raise errors.ClientErrorNotFound(
-                "{0} instance not found".format(self.type.__name__)
-            )
+            if hasattr(self.type, "__name__"):
```
Comment (demianbrecht) [inline django_declarative_apis/machinery/__init__.py:106]: Pedantic: EAFP > LBYL (try/except over if/else)

File: tests/models.py [hunk 6]
```diff
@@ -51,3 +51,14 @@ class ParentModel(models.Model):
 class RootNode(models.Model):
     id = models.IntegerField(primary_key=True)
     parent_field = models.ForeignKey(ParentModel, on_delete=models.CASCADE)
+
+
+try:
+    import dirtyfields
```
Comment (demianbrecht) [inline tests/models.py:57]: What's this about? `dirtyfields` is a hard requirement, no?

**Review by demianbrecht — COMMENTED** [review]
Do we care about preserving Django and Python backwards compatibility? Given we don't have an exhaustive automated test environment I'd say that we should limit the dependencies to what we actually use (and test) ourselves. Thoughts?

File: tests/models.py (hunk 6 above)
Comment (bgrant) [reply tests/models.py:57]: I was told that DDA _supports_ models that use `dirtyfields` but doesn't require it.  I don't know if that's true in practice.

**Comment by bgrant** [comment]
> Do we care about preserving Django and Python backwards compatibility? Given we don't have an exhaustive automated test environment I'd say that we should limit the dependencies to what we actually use (and test) ourselves. Thoughts?

Agreed.  I would ideally like to support versions of Python and Django that are currently "supported" by the maintainers of those projects, but that doesn't include Python 3.6 or Django 2.2 anymore.

File: django_declarative_apis/machinery/filtering.py [hunk 7]
```diff
@@ -9,6 +9,11 @@
 import inspect
 import types
 
+try:
+    from django.core.exceptions import FieldDoesNotExist
```
Comment (bgrant) [inline django_declarative_apis/machinery/filtering.py:13]: We can drop this too if we don't care about Django 2.2 compatibility:

https://docs.djangoproject.com/en/4.0/releases/3.1/#miscellaneous-1

**Review by demianbrecht — APPROVED** [review]

File: django_declarative_apis/machinery/__init__.py [hunk 8]
```diff
@@ -173,6 +177,35 @@ def __init__(cls, class_name, bases=None, dict=None):
...
+    dirty_dict = current_dirty_dict(resource)
+    resource_next, created = type(resource).objects.update_or_create(
+        pk=resource.pk, defaults=dirty_dict
+    )
+
+    # update fields in memory that changed on save to the database
+    field_name_to_att_name = {f.name: f.attname for f in resource._meta.concrete_fields}
+    for k, v in resource_next._as_dict(check_relationship=True).items():
+        att_key = field_name_to_att_name[k]
+        if getattr(resource, att_key, None) != v:
```
[the remark this answers is not in the document]
Comment (bgrant) [reply django_declarative_apis/machinery/__init__.py:201]: Yeah - this is @dshafer's code, but that's my guess.

**Pull request merged**